src/backend/renderer/utils.rs

```diff
@@ -112,6 +116,66 @@ impl RendererSurfaceState {
                 buffer_damage.dedup();
                 self.damage.push_front(buffer_damage);
                 self.damage.truncate(MAX_DAMAGE);
+
+                self.opaque_regions.clear();
+                if !self.buffer_has_alpha.unwrap_or(true) {
+                    self.opaque_regions.push(Rectangle::from_loc_and_size(
+                        (0, 0),
+                        self.surface_view.unwrap().dst,
+                    ))
+                } else if let Some(region_attributes) = &attrs.opaque_region {
+                    let opaque_regions = region_attributes
+                        .rects
+                        .iter()
+                        .map(|(kind, rect)| {
+                            let dest_size = self.surface_view.unwrap().dst;
+
+                            let rect_constrained_loc = rect
+                                .loc
+                                .constrain(Rectangle::from_extemities((0, 0), dest_size.to_point()));
+                            let rect_clamped_size = rect
+                                .size
+                                .clamp((0, 0), (dest_size.to_point() - rect_constrained_loc).to_size());
+
+                            let rect = Rectangle::from_loc_and_size(rect_constrained_loc, rect_clamped_size);
```

Comment:
But, if the rendering of a buffer from it's `src` to `dst` is scaled, shouldn't the opaque_region here also be scaled rather then cut of?

Reply:
I am not 100% sure, but I think the client should submit the correct size already.

The opaque region is defined as:
> The opaque region is specified in surface-local coordinates.

So it *should* be the same as the logical surface damage.